File: lecture_013_graph/L013.java
import java.util.ArrayList;
import java.util.LinkedList;

public class L013 {
    public static void main(String[] args) {
        solve();
    }

    public static void display(ArrayList<Integer>[] graph) {

        for (int i = 0; i < graph.length; i++) {
            System.out.print(i + "-> "); // u

            for (Integer e : graph[i]) {
                System.out.print(e + ", "); // v
            }
            System.out.println();
        }

        System.out.println();
    }

    public static void constructGraph(ArrayList<Integer>[] graph) {

        for (int i = 0; i < graph.length; i++) {
            graph[i] = new ArrayList<Integer>();
        }

        graph[7].add(4);
        graph[5].add(6);
        graph[5].add(0);
        // graph[0].add(5);

        graph[6].add(7);
        graph[7].add(6);
        graph[3].add(4);
        graph[2].add(3);
        graph[1].add(2);
        graph[1].add(0);

    }

    // topic.===============================================

    public static void solve() {
        int N = 8;
        ArrayList<Integer>[] graph = new ArrayList[N];
        ;
        constructGraph(graph);
        display(graph);
        dfs_improve(graph);
    }

    public static void topo_Sort(ArrayList<Integer>[] graph) {

        boolean[] vis = new boolean[graph.length];
        ArrayList<Integer> ans = new ArrayList<>();
        for (int i = 0; i < vis.length; i++) {
            if (!vis[i]) {
                topo_dfs(i, graph, vis, ans);
            }
        }

        for (Integer ele : ans) {
            System.out.println(ele);
        }

    }

    public static void topo_dfs(int src, ArrayList<Integer>[] graph, boolean[] vis, ArrayList<Integer> ans) {

        vis[src] = true;
        for (int ele : graph[src]) {
            if (!vis[ele]) {
                topo_dfs(ele, graph, vis, ans);
            }
        }

        ans.add(src);

    }

    public static void Kahn(ArrayList<Integer>[] graph) {

        int[] indegree = new int[graph.length];
        for (int i = 0; i < graph.length; i++) {
            for (int j = 0; j < graph[i].size(); j++)
                indegree[graph[i].get(j)]++;
        }

        LinkedList<Integer> q = new LinkedList<>();
        for (int i = 0; i < indegree.length; i++) {
            if (indegree[i] == 0)
                q.addLast(i);
        }

    }

    public static boolean dfs(int src, ArrayList<Integer>[] graph, boolean[] vis, boolean[] myPath) {

        if (vis[src]) {
            if (myPath[src])
                return true; // as in true for cycle exits
            return false;
        }

        vis[src] = myPath[src] = true;

        boolean res = false;

        for (int ele : graph[src]) {
            res = res || dfs(ele, graph, vis, myPath);
        }

        myPath[src] = false;
        return res;
    }

    public static void dfs_improve(ArrayList<Integer>[] graph) {

        int N = graph.length;
        boolean[] vis = new boolean[N];
        boolean[] myPath = new boolean[N];

        boolean res = false;
        for (int i = 0; i < graph.length; i++) {
            res = res || (dfs(i, graph, vis, myPath));
        }

        System.out.println(res);

    }

}

File: lecture_013_graph/L013.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <bits/stdc++.h>

using namespace std;

class Edge
{
public:
  int v;
  int w;

  Edge(int v1, int w1)
  {
    this->v = v1;
    this->w = w1;
  }
};

int N = 8;
vector<vector<Edge *>> graph(N, vector<Edge *>());

void addEdge(int u, int v, int w)
{
  graph[u].push_back(new Edge(v, w));
}

void display()
{

  for (int i = 0; i < N; i++)
  {
    cout << i << "->";
    for (int j = 0; j < graph[i].size(); j++)
    {
      Edge *e = graph[i][j];
      cout << "(" << e->v << "," << e->w << ")";
    }
    cout << endl;
  }
}

void construct()
{

  addEdge(0, 1, 10);
  addEdge(0, 2, 10);
  addEdge(1, 7, 10);
  addEdge(1, 6, 10);
  addEdge(1, 5, 10);
  addEdge(2, 6, 10);
  addEdge(3, 1, 10);
  addEdge(4, 2, 10);
  addEdge(4, 5, 10);
  addEdge(7, 1, 10);
  display();
}

vector<int> topodfs(int src, vector<bool> &vis, vector<int> &ans)
{
  vis[src] = true;

  for (int i = 0; i < graph[src].size(); i++)
  {
    Edge *e = graph[src][i];
    if (!vis[e->v])
      topodfs(e->v, vis, ans);
  }

  ans.push_back(src);
  return ans;
}

void topoSort()
{
  vector<bool> vis(N, false);
  vector<int> ret;
  vector<int> ans;

  for (int i = 0; i < N; i++)
  {
    if (!vis[i])
      ret = topodfs(i, vis, ans);
  }

  //print the ret in reverse
  for (int j = ans.size() - 1; j >= 0; j--)
    cout << ans[j] << ",";

  cout << endl;
}

void kahnalgo()
{
  //creation of indegree
  vector<int> indegree(N, 0);
  queue<int> q;
  vector<int> ans;

  for (int i = 0; i < N; i++)
  {
    for (int j = 0; j < graph[i].size(); j++)
      ++indegree[graph[i][j]->v];
  }

  for (int i = 0; i < N; i++)
  {
    if (indegree[i] == 0)
      q.push(i);
  }

  while (q.size() != 0)
  {
    int rvtx = q.front();
    q.pop();

    for (int i = 0; i < graph[rvtx].size(); i++)
    {
      Edge *e = graph[rvtx][i];
      indegree[e->v]--;
      if (indegree[e->v] == 0)
        q.push(e->v);
    }

    ans.push_back(rvtx);
  }

  if (ans.size() != N)
    cout << "Cycle" << endl;

  else
  {
    for (int j = 0; j < ans.size(); j++)
      cout << ans[j] << ",";
  }
}

bool cycledetectDFS(int src, vector<int> &vis, unordered_set<int> &ans)
{
  if (vis[src] == 1)
    return true;

  if (vis[src] == 2)
    return false;

  vis[src] = 1;
  bool res = false;

  for (int i = 0; i < graph[src].size(); i++)
  {
    res = res || cycledetectDFS(graph[src][i]->v, vis, ans);
  }

  vis[src] = 2;

  if (res)
    ans.insert(src);

  return res;
}

void cycledetect()
{
  vector<int> vis(N, -1);
  bool res = false;
  unordered_set<int> ans;
  for (int i = 0; i < N; i++)
  {
    if (vis[i] == -1)
      res = res || cycledetectDFS(i, vis, ans);
  }

  if (res)
    cout << "CYCLE \n";
  else
  {
    cout << "NO CYCLE \n";
    for (int i = ans.size() - 1; i >= 0; i--)
      cout << ans[i] << ",";
  }
}

void solve()
{

  construct();
  //topoSort();
  //kahnalgo();
  cycledetect();
}

int main()
{
  solve();
}


